### PR TITLE
Adding support for new 1.1 switchbot API version

### PR DESCRIPTION
**Pull request by edsaavedra84**
Also adding Lock device supprot with exact method entries

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,19 +12,41 @@ pip install python-switchbot
...
 ## Usage
 
-### Devices
+
+### API V1.1 Init
+```python
+from switchbot import SwitchBot
+
+import string
+import random
```
Comment (jonghwanhyeon) [inline README.md:21]:
```suggestion
import time
import hashlib
import hmac
import base64

from switchbot import SwitchBot
```

File: switchbot/__init__.py [hunk 2]
```diff
@@ -6,8 +6,8 @@
 
 
 class SwitchBot:
-    def __init__(self, token: str):
-        self.client = SwitchBotClient(token)
+    def __init__(self, token: str, ver: str = None, secret: str = None, nonce: str = None):
+        self.client = SwitchBotClient(token, ver, secret, nonce)
```
Comment (jonghwanhyeon) [inline switchbot/__init__.py]:
```suggestion
    def __init__(self, token: str, secret: str, nonce: str = ''):
        self.client = SwitchBotClient(token, secret, nonce)        
```

File: switchbot/client.py [hunk 3]
```diff
@@ -2,17 +2,46 @@
 
 import humps
 import requests
+import time
+import hashlib
+import hmac
+import base64
```
Comment (jonghwanhyeon) [inline switchbot/client.py:8]: import base64
import hashlib
import hmac
import time

import humps
import requests

Comment (jonghwanhyeon) [inline switchbot/client.py]:
```suggestion
import base64
import hashlib
import hmac
import time

import humps
import requests
```

File: switchbot/client.py [hunk 4]
```diff
@@ -2,17 +2,46 @@
...
 
-switchbot_host = 'https://api.switch-bot.com/v1.0'
-
+switchbot_hosts = {
+        '1.0': 'https://api.switch-bot.com/v1.0', 
+        '1.1': 'https://api.switch-bot.com/v1.1'
+    }
```
Comment (jonghwanhyeon) [inline switchbot/client.py]:
```suggestion
switchbot_host = 'https://api.switch-bot.com/v1.1'
```

File: switchbot/client.py [hunk 5]
```diff
@@ -2,17 +2,46 @@
...
 
 class SwitchBotClient:
-    def __init__(self, token: str):
+    def legacy_setup(self, token: str):
+        self.ver = "1.0"
+        self.session.headers['Authorization'] = token
+
+    def __init__(self, token: str, ver: str = None, secret: str = None, nonce: str = None):
```
Comment (jonghwanhyeon) [inline switchbot/client.py]:
```suggestion
    def __init__(self, token: str, secret: str, nonce: str = ''):
```

File: switchbot/client.py [hunk 6]
```diff
@@ -2,17 +2,46 @@
...
+        string_to_sign = bytes(string_to_sign, 'utf-8')
+
+        secret = bytes(secret, 'utf-8')
+
+        sign = base64.b64encode(hmac.new(secret, msg=string_to_sign, digestmod=hashlib.sha256).digest())
+
+        self.session.headers['sign'] = sign
         self.session.headers['Authorization'] = token
+        self.session.headers['nonce'] = nonce
+        self.session.headers['t'] = '{}'.format(t)
```
Comment (jonghwanhyeon) [inline switchbot/client.py]:
```suggestion

        timestamp = int(round(time.time() * 1000))
        string_to_sign = f'{token}{timestamp}{nonce}'

        string_to_sign = bytes(string_to_sign, 'utf-8')
        secret = bytes(secret, 'utf-8')
        sign = base64.b64encode(hmac.new(secret, msg=string_to_sign, digestmod=hashlib.sha256).digest())

        self.session.headers['Authorization'] = token
        self.session.headers['t'] = str(timestamp)
        self.session.headers['sign'] = sign
        self.session.headers['nonce'] = nonce
```

File: switchbot/client.py [hunk 7]
```diff
@@ -2,17 +2,46 @@
...
 
     def request(self, method: str, path: str, **kwargs) -> Any:
-        url = f'{switchbot_host}/{path}'
+        url = f'{switchbot_hosts[self.ver]}/{path}'
```
Comment (jonghwanhyeon) [inline switchbot/client.py]:
```suggestion
        url = f'{switchbot_host}/{path}'
```

File: setup.py [hunk 8]
```diff
@@ -5,7 +5,7 @@
 
 setup(
     name='python-switchbot',
-    version='2.1.0',
+    version='2.1.1',
```
Comment (jonghwanhyeon) [inline setup.py]:
```suggestion
    version='2.2.0',
```

File: README.md [hunk 9]
```diff
@@ -12,19 +12,41 @@ pip install python-switchbot
...
 ```python
 from switchbot import SwitchBot
 
 # To get the token, please refer to https://github.com/OpenWonderLabs/SwitchBotAPI#getting-started
 your_switch_bot_token = '98a6732b2ac256d40ffab7db31a82f518969f4d1a64eadff581d45e902327b7c577aa6ead517bda589c19b4ca0b2599b'
 switchbot = SwitchBot(token=your_switch_bot_token)
+```
+
+### Devices
+```python
```
Comment (jonghwanhyeon) [inline README.md]:
```suggestion

### Devices
```python
import uuid

from switchbot import SwitchBot

# To get the token and secret, please refer to https://github.com/OpenWonderLabs/SwitchBotAPI#getting-started
your_switch_bot_token = '98a6732b2ac256d40ffab7db31a82f518969f4d1a64eadff581d45e902327b7c577aa6ead517bda589c19b4ca0b2599b'
your_switch_bot_secret = '222cdc22f049d111c5d0071c131b8b77'
switchbot = SwitchBot(token=your_switch_bot_token, secret=your_switch_bot_secret, nonce=str(uuid.uuid4()))
```

Comment (jonghwanhyeon) [reply README.md]: According to the official document, they seem to suggest using uuid as nonce.

**Review by jonghwanhyeon — CHANGES_REQUESTED** [review]
Thanks for your contribution! Because SwitchBot API version has been officially updated to 1.1, I think it is better to support only that API and bump up the version to 2.2.0. Could you please check my suggestions?

**Comment by edsaavedra84** [comment]
> Thanks for your contribution! Because SwitchBot API version has been officially updated to 1.1, I think it is better to support only that API and bump up the version to 2.2.0. Could you please check my suggestions?

I think you are right and I agree with the proposed changes. 

More than glad to help!

**Comment by jonghwanhyeon** [comment]
Hello,

I have a question related to Issue #14 because I don't have SwitchBot Lock device 😢 .

In your pull request, you assigned `Lock.device_type_for` to `"Lock"`. However, according to [the official document](https://github.com/OpenWonderLabs/SwitchBotAPI#lock-1), it seems `"Smart Lock"` rather than `"Lock"`.

If possible, could you check which one is correct?

**Comment by edsaavedra84** [comment]
> Hello,
> 
> I have a question related to Issue #14 because I don't have SwitchBot Lock device 😢 .
> 
> In your pull request, you assigned `Lock.device_type_for` to `"Lock"`. However, according to [the official document](https://github.com/OpenWonderLabs/SwitchBotAPI#lock-1), it seems `"Smart Lock"` rather than `"Lock"`.
> 
> If possible, could you check which one is correct?

I am so sorry! I just saw the message.

It seems you already took care of that!

**Pull request merged**